herd/libdir/aarch64hwreqs.cat

```diff
@@ -56,104 +56,124 @@ let tc-ib = if "MTE" then [Imp & Tag & R]; iico_data; [B]; iico_ctrl; [((Exp & W
 let f-ib =
   [Imp & Instr & R]; iico_data; [B]; iico_ctrl; [_]
 
-(* HW TTD Updates permitted only for the AF or DB *)
-let HU = Imp & TTD & W
-assert empty HU \ (AF | DB)
-
 (* DSB-ordered-before *)
 let DSB-ob =
- [M | DC.CVAU | IC]; po; [dsb.full]; po; [~(Imp & M)]
+[M | DC.CVAU | IC]; po; [dsb.full]; po; [~(Imp & M) | Imp & Tag & R]
  | (if "ETS2" then [M | DC.CVAU | IC]; po; [dsb.full]; po; [Imp & TTD & M] else 0)
- | [(Exp & R) \ NoRet]; po; [dsb.ld]; po; [~(Imp & M)]
+ | [(Exp & R) \ NoRet]; po; [dsb.ld]; po; [~(Imp & M)| Imp & Tag & R]
  | (if "ETS2" then [(Exp & R) \ NoRet]; po; [dsb.ld]; po; [Imp & TTD & M] else 0)
- | [Exp & W]; po; [dsb.st]; po; [~(Imp & M)]
+ | [Exp & W]; po; [dsb.st]; po; [~(Imp & M) | Imp & Tag & R]
  | (if "ETS2" then [Exp & W]; po; [dsb.st]; po; [Imp & TTD & M] else 0)
 
 (* CSE-ordered-before *)
 let EXC-ENTRY-CSE = EXC-ENTRY
 let EXC-RET-CSE = if not "ExS" || "EOS" then EXC-RET else {}
 let CSE = ISB | EXC-ENTRY-CSE | EXC-RET-CSE
+
+let Sync-EXC = EXC-ENTRY | EXC-RET
+
 let CSE-ob =
  [Exp & R]; ctrl; [CSE]; po
+ | [Exp & R]; pick-ctrl-dep; [CSE]; po
+ | [Exp & R]; addr; [Exp & M]; po; [CSE]; po
+ | [Exp & R]; pick-addr-dep; [Exp & M]; po; [CSE]; po
  | DSB-ob; [CSE]; po
+ | [Imp & Instr & R]; f-ib; [CSE]; po
  | [Imp & TTD & R]; tr-ib; [CSE]; po
  | [Imp & TTD & R]; tr-ib; po; [CSE]; po
  | [Imp & Tag & R]; tc-ib; [CSE]; po
- | [Imp & Tag & R]; iico_data; [B]; iico_ctrl; [((Exp & M) \ Tag) | TagCheck & FAULT]; po; [CSE]; po
- | [Imp & Instr & R]; f-ib; [CSE]; po
+ | [Imp & Tag & R]; iico_data; [B]; iico_ctrl; [((Exp & M) \ Tag) | TagCheck & FAULT & Sync-EXC]; po; [CSE]; po
 
 (* Dependency-ordered-before *)
-let dob = addr 
-        | data
-        | ctrl; [W]
-        | addr; [Exp]; po; [W]
-        | addr; [Exp]; po; [CSE]; po; [R]
-        | addr; [Exp]; lrs
-        | data; [Exp]; lrs
+let dob =
+ addr; [Exp & M | Imp & Tag & R | Imp & TTD & R | HU | TLBI | DC.CVAU | IC]
```

Comment:
Isn't this all of addr? Is anything excluded?

Reply:
Yes, this is all of `addr` as currently defined.

Reply:
I suppose, the list of endpoint constraints could be moved into the definition of `addr` itself, if that helps.

Reply:
Maybe that would improve the readability of the graphs too, what do you think? Whichever option we choose here should apply for data and ctrl.

Reply:
An advantage of moving the endpoint constraints for `addr`, `data` and `ctrl` closer to their definitions is that the next time the definition extends to new events, there is one place to edit it, not both the definition and the `dob` like now -- which is a readability/maintainability win for the transliteration of the rules.

The disadvantage is that we do need to define dependencies more generally, and use them in more specific ways:
* for `ctrl`, the `dob` does not provide order to`CSE` effects, but we do use `ctrl;[CSE];po` elsewhere.
* for `pick-addr-dep`, the `pob` does not provide order to `Exp & R`, but `pick-addr-dep; [Exp & M]; po; [Exp & W]` makes use of it.

Hence, to preserve the same behaviour we'd need to:
* Move the endpoint constraints for `ctrl` into its definition, extend them with `CSE`, but let `ctrl; [~CSE]` in `dob`. Admittedly, there are no tests that tell the difference between having or not having `CSE` there.
* Move the endpoint constraints for `pick-addr-dep` into its definition, extend them with `Exp & R`, but let `pick-addr-dep; [~(Exp & R)]` in `pob`. The need for this is illustrated by `catalogue/aarch64-pick/tests/T3-bis.litmus`, which is forbidden.